src/tests/test_frodokem.cpp

```diff
@@ -0,0 +1,224 @@
+   };
+
+   std::vector<Test::Result> results;
+   for(auto mode : modes) {
+      Botan::FrodoKEMMode m(mode);
+      Botan::FrodoKEMConstants consts(mode);
+      Test::Result& result = results.emplace_back("FrodoKEM roundtrip: " + m.to_string());
+
+      Botan::FrodoKEM_PrivateKey sk1(rng, mode);
+      Botan::FrodoKEM_PublicKey pk1(sk1.public_key_bits(), mode);
+
+      // Happy case
+      Botan::PK_KEM_Encryptor enc1(pk1, "Raw");
+      const auto enc_res = enc1.encrypt(rng, 0 /* no KDF */);
+
+      result.test_eq("length of shared secret", enc_res.shared_key().size(), enc1.shared_key_length(0));
+      result.test_eq("length of ciphertext", enc_res.encapsulated_shared_key().size(), enc1.encapsulated_key_length());
+
+      Botan::PK_KEM_Decryptor dec1(sk1, rng, "Raw");
+      auto ss = dec1.decrypt(enc_res.encapsulated_shared_key(), 0 /* no KDF */);
+
+      result.test_eq("shared secrets match", ss, enc_res.shared_key());
+      result.test_eq("length of shared secret (decaps)", ss.size(), dec1.shared_key_length(0));
+
+      // Decryption failures ("All right then, keep your secrets.")
+      Botan::FrodoKEM_PrivateKey sk2(rng, mode);
+
+      // Decryption failure: mismatching private key
+      Botan::PK_KEM_Decryptor dec2(sk2, rng, "Raw");
+      auto ss_mismatch = dec2.decrypt(enc_res.encapsulated_shared_key(), 0 /* no KDF */);
+      result.test_eq("decryption failure sk",
+                     ss_mismatch,
+                     get_decryption_error_value(consts, enc_res.encapsulated_shared_key(), sk2));
+
+      // Decryption failure: bitflip in encapsulated shared value
+      const auto mutated_encaps_value = Test::mutate_vec(enc_res.encapsulated_shared_key());
+      ss_mismatch = dec2.decrypt(mutated_encaps_value, 0 /* no KDF */);
+      result.test_eq(
+         "decryption failure bitflip", ss_mismatch, get_decryption_error_value(consts, mutated_encaps_value, sk2));
+
+      // Decryption failure: malformed encapsulation value
+      result.test_throws(
+         "malformed encapsulation value", "FrodoKEM ciphertext does not have the correct byte count", [&] {
+            auto short_encaps_value = enc_res.encapsulated_shared_key();
+            short_encaps_value.pop_back();
+            dec1.decrypt(short_encaps_value, 0);
+         });
+   }
+
+   return results;
+}
+
+class Frodo_Keygen_Tests final : public PK_Key_Generation_Test {
+   public:
+      std::vector<std::string> keygen_params() const override {
+         return {
+            "FrodoKEM-640-SHAKE", "FrodoKEM-976-SHAKE", "eFrodoKEM-640-SHAKE", "eFrodoKEM-976-SHAKE",
+   #if defined(BOTAN_HAS_AES)
+               "FrodoKEM-640-AES", "FrodoKEM-976-AES", "eFrodoKEM-640-AES", "eFrodoKEM-976-AES",
+   #endif
+         };
+      }
+
+      std::string algo_name() const override { return "FrodoKEM"; }
+};
+
+}  // namespace
+
+   #if defined(BOTAN_HAS_AES)
+BOTAN_REGISTER_TEST("frodokem", "frodo_kat_tests", Frodo_KAT_Tests);
+   #endif
```

Comment:
We don't run any of the KATs if AES is unavailable??? That puts us in a situation where we are willing to configure and build something and then not test it at all :(

Reply:
As with elsewhere we should just skip AES tests at runtime if we don't have support.

Reply:
done.

Reply:
There's a distinction between (not) having FrodoKEM-AES and not having AES, since the KATs use AES for seed expansion even for FrodoKEM-SHAKE.

With the new FrodoKEM-AES module we can make the following minimized builds:

- `--enable-modules=frodokem_aes` (Can run FrodoKEM-SHAKE/-AES and all FrodoKEM-SHAKE/-AES tests)
- `--enable-modules=aes,frodokem` (Can run FrodoKEM-SHAKE and all FrodoKEM-SHAKE tests)
-  `--enable-modules=frodokem` (Can run FrodoKEM-SHAKE but not test its KATs, this does not compile right now because we have to use `CTR_DRBG_AES256` in the KATs, which was previously only compiled if `BOTAN_HAS_AES`)

I believe the story is similar in the Kyber implementation, where we skip the KATs  if we do not have AES available.

A possible fix is to manually enter the expanded seed values into the test vector file but I'm not sure if we want to do that or just skip like before/in Kyber?

Reply:
@atreiber94 is right. We skip KATs because they depend on the test-helper `CTR_DRBG_AES256` that depends on AES. 
Meanwhile, I fixed the compilation error, @atreiber94 pointed out. By disabling the KAT tests for builds w/o AES. 😭
Note, though, that at least we do roundtrip tests with FrodoKEM if the KATs are disabled. Obviously, that's then not hinged to an independent test vector.

In any case, other PQC algorithms currently have the very same problem. I would suggest addressing this independent of this pull request. For instance, we could create a reduced set of KATs that don't rely on `CTR_DRBG_AES256` to at least have an independent sanity check.

Additionally, and somewhat orthogonally, I suggest to add a generic KAT test class (along the lines of the base classes in `src/tests/test_pubkey.h`) for all PQC-KATs to share. This could replace the individual roundtrip and encoding/decoding tests that we currently have for each algorithm.

Reply:
> I suggest to add a generic KAT test class (along the lines of the base classes in `src/tests/test_pubkey.h`) for all PQC-KATs to share.

Done in #3807.

Reply:
With #3807 merged, I closed #3808 and incorporated the relevant adaptions into this pull request.